localization/localization_data.py

```diff
@@ -0,0 +1,243 @@
+print("Found %d points well separated" % len(centers_guess))
+
+# ###############################
+# do localization
+# ###############################
+# roi sizes
+xy_roi_size = 12 * sigma_xy
+z_roi_size = 3 * sigma_z
+
+print("fitting %d rois" % centers_guess.shape[0])
+results = joblib.Parallel(n_jobs=-1, verbose=10, timeout=None)(
+    joblib.delayed(localize.fit_roi)(centers_guess[ii], imgs_filtered, dc, theta, x, y, z, xy_roi_size, z_roi_size, 0.2, 3) for ii in range(len(centers_guess)))
+# for ii in range(len(centers_guess)):
+#     localize.fit_roi(centers_guess[ii], imgs_filtered, dc, theta, x, y, z, xy_roi_size, z_roi_size, 0.2, 3)
+
+# # unpack results
+fit_params, ntries, rois, centers_fit_sequence = list(zip(*results))
+fit_params = np.asarray(fit_params)
+ntries = np.asarray(ntries)
+rois = np.asarray(rois)
+centers_fit_sequence = np.asarray(centers_fit_sequence)
+centers_fit = np.concatenate((fit_params[:, 3][:, None], fit_params[:, 2][:, None], fit_params[:, 1][:, None]), axis=1)
+
+# only keep points if think the fit was good
+# first check center is within our region
+to_keep_x = np.logical_and(centers_fit[:, 2] >= x.min(), centers_fit[:, 2] <= x.max())
+to_keep_yz = np.logical_and(centers_fit[:, 0] <= centers_fit[:, 1] * np.tan(theta),
+                            centers_fit[:, 0] >= (centers_fit[:, 1] - dstage * npos) * np.tan(theta))
+to_keep_z = np.logical_and(centers_fit[:, 0] >= 0, centers_fit[:, 0] <= ny * dc * np.sin(theta))
+
+to_keep_position = np.logical_and(np.logical_and(to_keep_x, to_keep_yz), to_keep_z)
+
+# check that size was reasonable
+to_keep_sigma_xy = np.logical_and(fit_params[:, 4] <= sigma_xy*3, fit_params[:, 4] >= 0.25 * sigma_xy)
+to_keep_sigma_z = np.logical_and(fit_params[:, 5] <= sigma_z*2, fit_params[:, 5] >= 0.25 * sigma_z)
+
+to_keep = np.logical_and(np.logical_and(to_keep_sigma_xy, to_keep_sigma_z), to_keep_position)
+
+# plot localization fit diagnostic on good points
+# plt.ioff()
+# plt.switch_backend("agg")
+# print("plotting %d ROI's" % np.sum(to_keep))
+# inds = np.arange(len(to_keep))[to_keep]
+# results = joblib.Parallel(n_jobs=-1, verbose=10, timeout=None)(
+#     joblib.delayed(plot_roi)(fit_params[ii], rois[ii], imgs_filtered, theta, x, y, z,
+#                               center_guess=centers_fit_sequence[ii], figsize=figsize,
+#                               prefix=("%04d" % ii), save_dir=save_dir)
+#     for ii in inds)
+
+# plt.ion()
+# # plt.switch_backend("TkAgg")
+# plt.switch_backend("Qt5Agg")
+
+
+# only keep unique center if close enough
+centers_unique = localize.combine_nearby_peaks(centers_fit[to_keep], min_xy_dist, min_z_dist, mode="keep-one")
+print("identified %d unique points" % len(centers_unique))
+
+# ###############################
+# interpolate images so are on grids in coverslip coordinate system and plot all results
+# ###############################
+xi, yi, zi, imgs_unskew = localize.interp_opm_data(imgs_filtered, dc, dstage, theta, mode="ortho-interp")
```

Comment:
Out of curiosity, is there a difference between the orthogonal interpolation that you wrote and that I wrote?

Does it make sense to refactor and get OPM specific functions into their own library?

Reply:
@dpshepherd mine was just a way to build my understanding. I am using it in some of the localization tests, but ultimately it should be removed.

I agree it might make sense to put any OPM specific stuff in a separate place than localization.py

Right now the organization is not great. I pushed the change as I started trying to localize the full image and wanted to work on a different computer

Reply:
Sounds good. Let's brainstorm how to better organize, I agree what I have is all over the place. Things on my end have become really messy recently with all of the changes.

Reply:
I just meant the stuff I pushed was disorganized